### PR TITLE
[example] Add implicit fem example

### PR DESCRIPTION
**Pull request by BillXu2000**
An implicit fem example, and also an example to show the usage of dynamic indexing.
Notes for reviewer:
1. Some kernels might be too magic or not fully compatible with Taichi syntax concept.
2. `_ti_core.GGUI_AVAILABLE` is necessary for ggui environment detection, but I am not sure whether it is safe to use it in example for long.

### CODE REVIEW

File: python/taichi/examples/simulation/implicit_fem.py [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+        for i in range(10):
+            get_force()
+            advect()
+    else:
+        for i in range(1):
+            cg()
+    floor_bound()
+
+
+get_vertices()
```
Comment (k-ye) [inline python/taichi/examples/simulation/implicit_fem.py:285]: nit: could you `if __name__ == '__main__':` ?

File: python/taichi/examples/simulation/implicit_fem.py [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+                                 movement_speed=0.03,
+                                 hold_key=ti.ui.RMB)
+        scene.set_camera(camera)
+
+        scene.ambient_light((0.1, ) * 3)
+
+        scene.point_light(pos=(0.5, 10.0, 0.5), color=(0.5, 0.5, 0.5))
+        scene.point_light(pos=(10.0, 10.0, 10.0), color=(0.5, 0.5, 0.5))
+
+        #scene.particles(x, radius=0.02)
```
Comment (strongoier) [inline python/taichi/examples/simulation/implicit_fem.py:312]: nit: we'd better not leave commented code

File: python/taichi/examples/simulation/implicit_fem.py [hunk 3]
```diff
@@ -0,0 +1,353 @@
+import argparse
+
+import numpy as np
+from taichi._lib import core as _ti_core
+
+import taichi as ti
+
+parser = argparse.ArgumentParser()
+parser.add_argument('--exp', default='implicit')
+#parser.add_argument('--exp', default='explicit')
```
Comment (strongoier) [inline python/taichi/examples/simulation/implicit_fem.py:10]: ditto

**Review by strongoier — APPROVED** [review]
Looks great! Thanks!

BTW using `_ti_core.GGUI_AVAILABLE` is OK for now. We will design & add new APIs for detecting such things in the future.

File: python/taichi/examples/simulation/implicit_fem.py (hunk 3 above)
Comment (BillXu2000) [reply python/taichi/examples/simulation/implicit_fem.py:10]: User can comment line 9 and uncomment line 10 to switch to the explicit version, so I think this line can be kept here.
Same for line 13 and 14.

File: python/taichi/examples/simulation/implicit_fem.py (hunk 2 above)
Comment (BillXu2000) [reply python/taichi/examples/simulation/implicit_fem.py:312]: I will fix this.

File: python/taichi/examples/simulation/implicit_fem.py (hunk 1 above)
Comment (BillXu2000) [reply python/taichi/examples/simulation/implicit_fem.py:285]: I will fix this.

File: python/taichi/examples/simulation/implicit_fem.py (hunk 3 above)
Comment (k-ye) [reply python/taichi/examples/simulation/implicit_fem.py]: FYI you can use `choices`: https://docs.python.org/3/library/argparse.html#choices

Comment (BillXu2000) [reply python/taichi/examples/simulation/implicit_fem.py]: I will use `choices`.

**Pull request merged**